pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionsStatsGenerator.java

```diff
@@ -0,0 +1,93 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.functions.worker;
+
+import org.apache.pulsar.functions.proto.InstanceCommunication;
+import org.apache.pulsar.functions.runtime.Runtime;
+import org.apache.pulsar.functions.runtime.RuntimeSpawner;
+import org.apache.pulsar.common.util.SimpleTextOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * A class to generate stats for pulsar functions running on this broker
+ */
+public class FunctionsStatsGenerator {
+
+    private static final Logger log = LoggerFactory.getLogger(FunctionsStatsGenerator.class);
+
+    public static void generate(WorkerService workerService, String cluster, SimpleTextOutputStream out) {
+        if (workerService != null) {
+            Map<String, FunctionRuntimeInfo> functionRuntimes
+                    = workerService.getFunctionRuntimeManager().getFunctionRuntimeInfos();
+
+            for (Map.Entry<String, FunctionRuntimeInfo> entry : functionRuntimes.entrySet()) {
+                String fullyQualifiedInstanceName = entry.getKey();
+                FunctionRuntimeInfo functionRuntimeInfo = entry.getValue();
+                RuntimeSpawner functionRuntimeSpawner = functionRuntimeInfo.getRuntimeSpawner();
+
+                if (functionRuntimeSpawner != null) {
+                    Runtime functionRuntime = functionRuntimeSpawner.getRuntime();
+                    if (functionRuntime != null) {
+                        try {
+                            InstanceCommunication.MetricsData metrics = functionRuntime.getAndResetMetrics().get();
+                            for (Map.Entry<String, InstanceCommunication.MetricsData.DataDigest> metricsEntry
+                                    : metrics.getMetricsMap().entrySet()) {
+                                String metricName = metricsEntry.getKey();
+                                InstanceCommunication.MetricsData.DataDigest dataDigest = metricsEntry.getValue();
+
+                                String tenant = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getTenant();
+                                String namespace = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getNamespace();
+                                String name = functionRuntimeInfo.getFunctionInstance()
+                                        .getFunctionMetaData().getFunctionConfig().getName();
+                                int instanceId = functionRuntimeInfo.getFunctionInstance().getInstanceId();
```

Comment:
We should have some test to verify the metrics output. There is already a parser for the Prometheus format in the tests code. It parses the stats into a map. Take a look at `PrometheusMetricsTest`

Reply:
sounds good. I will add a test